src/pylibsshext/sftp.pyx

```diff
@@ -54,69 +57,186 @@ cdef class SFTP:
             remote_file_b = remote_file.encode("utf-8")
 
-        rf = sftp.sftp_open(self._libssh_sftp_session, remote_file_b, O_RDONLY, sftp.S_IRWXU)
-        if rf is NULL:
-            raise LibsshSFTPException("Opening remote file [%s] for read failed with error [%s]" % (remote_file, self._get_sftp_error_str()))
-
-        while True:
-            file_data = sftp.sftp_read(rf, <void *>read_buffer, sizeof(char) * 1024)
-            if file_data == 0:
-                break
-            elif file_data < 0:
-                sftp.sftp_close(rf)
-                raise LibsshSFTPException("Reading data from remote file [%s] failed with error [%s]"
-                                          % (remote_file, self._get_sftp_error_str()))
-
-            with open(local_file, 'wb+') as f:
-                bytes_written = f.write(read_buffer[:file_data])
-                if bytes_written and file_data != bytes_written:
-                    sftp.sftp_close(rf)
-                    raise LibsshSFTPException("Number of bytes [%s] read from remote file [%s]"
-                                              " does not match number of bytes [%s] written to local file [%s]"
-                                              " due to error [%s]"
-                                              % (file_data, remote_file, bytes_written, local_file, self._get_sftp_error_str()))
-        sftp.sftp_close(rf)
+        attrs = sftp.sftp_stat(self._sftp, remote_file_b)
+        if attrs is NULL:
+            raise LibsshSFTPException("Failed to stat the remote file [%s] with error [%s]"
+                                      % (remote_file, self._get_sftp_error_str()))
+        self._file_size = attrs.size
 
-    def close(self):
-        if self._libssh_sftp_session is not NULL:
-            sftp.sftp_free(self._libssh_sftp_session)
-            self._libssh_sftp_session = NULL
+        buffer_size = min(self._limits.max_read_length, self._file_size)
+        try:
+            buffer = <char *>PyMem_Malloc(buffer_size)
 
-    def _get_sftp_error_str(self):
-        error = sftp.sftp_get_error(self._libssh_sftp_session)
-        if error in MSG_MAP and error != sftp.SSH_FX_FAILURE:
-            return MSG_MAP[error]
-        return "Generic failure: %s" % self.session._get_session_error_str()
+            rf = sftp.sftp_open(self._sftp, remote_file_b, O_RDONLY, sftp.S_IRWXU)
+            if rf is NULL:
+                raise LibsshSFTPException("Opening remote file [%s] for reading failed with error [%s]" % (remote_file, self._get_sftp_error_str()))
+            self._rf = rf
+
+            with open(local_file, 'wb') as f:
+                # open up to 10 parallel transfers
+                i = 0
+                while i < 10 and self._total_bytes_requested < self._file_size:
+                    self._get_chunk()
+                    i += 1
+
+                while len(self._aio_queue):
+                    aio = self._aio_queue.pop()
```

Comment:
Similar to the previous comment for pop() in case of write.

From what I understand, in case of pop, assuming requests Rq_1, Rq_2 ... Rq_10 are issued:

In case of download, waiting for Rq_10's response first (due to pop) would write the data of response 10 first in the local file before writing the data that comes as request 1's response, which isn't the desired order for file contents.

Hence, I think the usage of pop here should also be changed in a similar way as suggested for write.

Reply:
Thanks! Fixed now!